### PR TITLE
Support gradients for slide background

### PR DESCRIPTION
**Pull request by tjinauyeung**
Adds gradient support for slide backgrounds.

**Usage**
```js
slide.background = { 
  gradient: { 
    angle: 90, 
    stops: { 
      0: "#FF00FF", 
      100: "#00FF00" 
    } 
  } 
}
```

### How to test
1. In your terminal, go to demo folder `cd ./demos`
2. Run `npm i`
3. Run `node ./browser_server.mjs`
4. Open up `http://localhost:8000/browser/index.html` in the browser
5. Replace the html snippet and run export

### CODE REVIEW

File: src/gen-utils.ts [hunk 1]
```diff
@@ -198,13 +198,39 @@ export function createGlowElement(options: TextGlowProps, defaults: TextGlowProp
...
+ * Create gradient list
+ * @param {Gradient} gradient
+ * @param {string} internalElements
+ * @returns XML string
+ */
+function createGradientList(gradient: Gradient, internalElements: string): string {
+	const multiplier = 1000
+	const gradientStops = Object.keys(gradient)
+		.filter(key => Number.isFinite(Number(key)))
+		.sort()
```
Comment (tjinauyeung) [inline src/gen-utils.ts]: does it need to be sorted?

File: src/gen-utils.ts [hunk 2]
```diff
@@ -198,13 +198,39 @@ export function createGlowElement(options: TextGlowProps, defaults: TextGlowProp
...
+		.map(pos => `<a:gs pos="${Number(pos) * multiplier}">${createColorElement(gradient[pos], internalElements)}</a:gs>`)
+	return `<a:gsLst>${gradientStops.join('')}</a:gsLst>`
+}
+
+/**
+ * Create linear gradient
+ * @param {number} angle
+ * @returns XML string
+ */
+function createLinearGradient(angle = 90): string {
```
Comment (tjinauyeung) [inline src/gen-utils.ts]: default to 0?

File: src/gen-utils.ts [hunk 3]
```diff
@@ -198,13 +198,39 @@ export function createGlowElement(options: TextGlowProps, defaults: TextGlowProp
...
+}
+
 /**
  * Create color selection
  * @param {Color | ShapeFillProps | ShapeLineProps} props fill props
  * @returns XML string
  */
 export function genXmlColorSelection(props: Color | ShapeFillProps | ShapeLineProps): string {
 	let fillType = 'solid'
+	let gradient = {} as Gradient
```
Comment (tjinauyeung) [inline src/gen-utils.ts]: avoid as

File: src/slide.ts [hunk 4]
```diff
@@ -95,6 +95,7 @@ export default class Slide {
 	 * @type {BackgroundProps}
 	 * @example solid color `background: { color:'FF0000' }`
 	 * @example color+trans `background: { color:'FF0000', transparency:0.5 }`
+	 * @example gradient `background: { gradient: { 0:'FF0000', 100:'00FFFF', angle: 45 }`
```
Comment (tjinauyeung) [inline src/slide.ts]: fmt

File: src/core-interfaces.ts [hunk 5]
```diff
@@ -176,6 +181,14 @@ export interface ShapeFillProps {
 	 * @example pptx.SchemeColor.text1 // Theme color (Text1)
 	 */
 	color?: Color
+	/**
+	 * Gradient color
+	 * - Key inside object is the color position in percentage
+	 * - `HextColor` or `ThemeColor`
+	 * @example { 0: 'FF0000', 100: '00FF00' }
```
Comment (tjinauyeung) [inline src/core-interfaces.ts]: add angle

File: src/gen-utils.ts [hunk 6]
```diff
@@ -198,6 +198,67 @@ export function createGlowElement(options: TextGlowProps, defaults: TextGlowProp
...
 
+/**
+ * Create gradient elements i.e. `a:gsLst` and `a:lin`
+ * @param {Gradient} gradient
+ * @param {string} internalElements
+ * @see http://officeopenxml.com/drwSp-GradFill.php
+ * @returns XML string
+ */
+function createGradientElements(gradient: Gradient, internalElements: string): string {
+	let strXml = '',
```
Comment (tjinauyeung) [inline src/gen-utils.ts:209]: Very imperative code here, but consistent with the style used in the project, see `createGlowElement` above

File: src/gen-utils.ts [hunk 7]
```diff
@@ -198,6 +198,66 @@ export function createGlowElement(options: TextGlowProps, defaults: TextGlowProp
...
+		stops = gradient.stops
+
+	if (!isGradient(gradient)) {
+		console.warn(`This is not a valid gradient:`, gradient)
```
Comment (tjinauyeung) [inline src/gen-utils.ts:213]: This is handled similar to colors.

When no valid color is found, we log a warning in the console, and use a default color. 

See `createColorElement` fn.

**Comment by sam-pitch** [comment]
~~When trying to run the `node ./browser_server.mjs` I get~~
I just had to run `npm i` first.
```bash
internal/process/esm_loader.js:74
    internalBinding('errors').triggerUncaughtException(
                              ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'express' imported from /home/sam/Code/repos/PptxGenJS/demos/browser_server.mjs
    at new NodeError (internal/errors.js:322:7)
    at packageResolve (internal/modules/esm/resolve.js:732:9)
    at moduleResolve (internal/modules/esm/resolve.js:773:18)
    at Loader.defaultResolve [as _resolve] (internal/modules/esm/resolve.js:887:11)
    at Loader.resolve (internal/modules/esm/loader.js:89:40)
    at Loader.getModuleJob (internal/modules/esm/loader.js:242:28)
    at ModuleWrap.<anonymous> (internal/modules/esm/module_job.js:76:40)
    at link (internal/modules/esm/module_job.js:75:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```

**Comment by sam-pitch** [comment]
Screenshot from 2022-07-27 09-44-05

Looks very good to me!

**Review by sam-pitch — APPROVED** [review]

**Review by zarkone — APPROVED** [review]
looks great! 

Interesting that all the compiled code is not in .gitignore

**Review by pastafari — APPROVED** [review]

**Comment by pastafari** [comment]
@tjinauyeung can we make a plan for attempting to merge changes to upstream? 
e.g
- Should we make a PR from `add-gradients` to upstream/master or from `pitch-main` to upstream/master?
- How do we rebase against changes to upstream?

**Pull request merged**